### PR TITLE
Support Stim DEM strings in get_decoder

### PR DESCRIPTION
**Pull request by vedika-saravanan**
## Description

Adds unified decoder initialization from either a parity-check matrix or raw Stim detector error model text.

This supports both existing PCM-based decoders and DEM-native decoders needed by #546. PCM-based decoders can parse DEM text into `H`, `O`, and `error_rate_vec` defaults, while DEM-native decoders can consume the raw DEM text without going through a lossy matrix conversion.

### API (`cudaq/qec/decoder.h`)

- Adds `decoder_init = std::variant<sparse_binary_matrix, std::string>` as the decoder registry construction input.
- Keeps existing PCM construction paths:
  - `get_decoder(name, H, options)`
  - `decoder::get(name, H, options)`
- Adds DEM string construction through the same entry points:
  - `get_decoder(name, dem_text, options)`
  - `decoder::get(name, dem_text, options)`
- Adds `dem_from_stim_text(dem_text)` to parse Stim DEM text into `detector_error_model`.
- Adds helper routing for PCM-based decoders so DEM-derived `O` and `error_rate_vec` are supplied as defaults when not explicitly provided by the user.

The DEM-to-detector_error_model parser is lossy: it extracts detector flips, observable flips, and per-error probabilities, but drops detector coordinates and separator-encoded correlation structure. DEM-native decoders should consume the raw string alternative in decoder_init.

Python: `cudaq_qec.get_decoder(...)` now accepts Stim DEM strings. Python-registered decoders currently receive parsed `H` plus DEM-derived `O` / `error_rate_vec` defaults, not raw DEM text.

### Dependency / build

QEC now provides its own Stim dependency via `FetchContent` when `libstim` is not already available, so standalone QEC builds do not depend on parent CUDA-Q interim build artifacts.

### Tests

C++ gtests and Python tests cover:

- DEM string construction through `get_decoder(...)`
- PCM construction still working
- user-provided options overriding DEM-derived defaults
- DEM parse edge cases
- DEMs without observables
- `stim::DemTarget` category assumptions

### Out of scope / follow-ups

Chromobius plugin integration itself; detector-coordinate storage on `detector_error_model`; optional PyMatching-specific improvements; user-facing Sphinx/RST docs.

## Runtime / performance impact

N/A

## Self-review checklist

Please confirm each item before requesting review. Check `[x]` or strike
through and explain.

### Before requesting review
- [x] I reviewed my own full diff in GitHub or my editor.
- [x] PR is in Draft if it is not yet ready for review.
- [x] Temporary / debugging changes have been removed.
- [x] Local test logs reviewed; no unexplained warnings or errors.
- [x] CI logs reviewed; no unexplained warnings or errors.
- [x] Full CI has been run.

### Scope and size
- [x] PR is under ~1000 lines, or an exception is justified in the description.
- [x] Refactoring-only changes are isolated in their own PR(s).
- [x] No existing tests were disabled or modified just to make this PR pass
      (if so, an issue has been raised).

### Tests
- [x] New functionality has new tests.
- [x] Tests fail if the new functionality is broken (including crashes), not
      just when it is missing.
- [x] Negative tests added where exceptions are expected.
- [x] Truth data added where simple `EXPECT_*` / `assert` checks are
      insufficient for algorithmic correctness.
- [x] CI runtime impact considered; team notified if significant.

### Documentation
- [x] Public-facing APIs have Doxygen docs.
- [x] User-visible behavior changes have public docs, or a follow-up is
      tracked.
- [x] User-facing docs for new features are in a **separate PR** held until
      release (the docs site publishes immediately on merge to the default
      branch, so feature docs must not land before the feature ships).

### Code style
- [x] Naming follows the existing convention (`snake_case` vs `camelCase`) for
      the area being modified.

### Dependencies
- [x] No new third-party dependencies, **or** the team has been notified and
      OSRB tickets filed.

### CODE REVIEW

File: libs/qec/include/cudaq/qec/decoder.h [hunk 1]
```diff
@@ -427,4 +430,38 @@ inline void convert_vec_hard_to_soft(const std::vector<std::vector<t_hard>> &in,
 std::unique_ptr<decoder>
 get_decoder(const std::string &name, const cudaqx::tensor<uint8_t> &H,
             const cudaqx::heterogeneous_map options = {});
+
+struct detector_error_model;
+
+/// @brief DEM-derived defaults; pointers alias into the source `dem`.
+struct dem_default_values {
+  const cudaqx::tensor<uint8_t> *O = nullptr;
```
Comment (bmhowe23) [inline libs/qec/include/cudaq/qec/decoder.h:524]: I'm open to feedback, but this makes me think that we should merge #550 first and then update these changes to use the new type before merging this PR. Otherwise, we might forget parts like this that need to change. Thoughts?

File: libs/qec/lib/CMakeLists.txt [hunk 2]
```diff
@@ -60,6 +61,16 @@ add_library(${LIBRARY_NAME} SHARED ${QEC_SOURCES})
 add_subdirectory(decoders/plugins/example)
 add_subdirectory(decoders/plugins/pymatching)
 
+# libstim comes from the parent build (CUDA-Q).
```
Comment (bmhowe23) [inline libs/qec/lib/CMakeLists.txt]: Perhaps I'm missing something, but I'm a bit confused how this is working because I didn't think that CUDA-Q was redistributing `libstim`. Is it possible `libstim` is coming from somewhere other than the parent CUDA-Q build?

Comment (vedika-saravanan) [reply libs/qec/lib/CMakeLists.txt]: libstim is not redistributed as a standalone CUDA-Q library. CUDA-Q vendors Stim as a submodule and adds it to the same CMake build via `add_subdirectory(tpls/Stim ...)`, which creates the upstream libstim target. This QEC target is just consuming that in-tree target during the combined build and linking it privately/static into libcudaq-qec.

I’ll update the comment to make that clearer, e.g. “libstim is defined by the vendored Stim submodule added by the parent CUDA-Q build.”

File: libs/qec/include/cudaq/qec/decoder.h (hunk 1 above)
Comment (vedika-saravanan) [reply libs/qec/include/cudaq/qec/decoder.h:524]: I agree #550 should land first because it builds on the dense tensor<uint8_t> decoder signature that #550 swaps out for sparse_binary_matrix. Probably cleanest if #550 goes in first and I rebase this on top.

File: libs/qec/lib/CMakeLists.txt (hunk 2 above)
Comment (bmhowe23) [reply libs/qec/lib/CMakeLists.txt]: OK, that makes sense, but I think we should be able to build in a system or container without requiring interim build artifacts from CUDA-Q. (For example, we should be able to build using the nightly CUDA-Q image, which does not have the interim build artifacts in it.) Could you please just add our own Stim via FetchContent in our CMakeLists.txt files?

File: libs/qec/include/cudaq/qec/decoder.h [hunk 3]
```diff
@@ -427,4 +430,38 @@ inline void convert_vec_hard_to_soft(const std::vector<std::vector<t_hard>> &in,
...
+/// User-supplied values for either key win over the DEM-derived ones.
+///
+/// @note Lossy: detector annotations, decomposition separators, and
+/// `error_ids` are dropped. Sufficient for matching-style / H-based
+/// decoders (LUT, NV, sliding_window, TRT, PyMatching). Decoders that
+/// need full DEM metadata (e.g. Chromobius detector color/basis) require
+/// the planned \c detector_coords extension on
+/// \c cudaq::qec::detector_error_model; tracked as a follow-up.
+std::unique_ptr<decoder>
+get_decoder_from_stim_dem(const std::string &name,
```
Comment (bmhowe23) [inline libs/qec/include/cudaq/qec/decoder.h]: Per our offline conversation, I think this function could be removed if we change the decoder extension point / decoder constructors to accept `std::variant<tensor, const char *>` or `std::variant<tensor, string_view>`.

File: libs/qec/lib/CMakeLists.txt [hunk 4]
```diff
@@ -58,6 +60,25 @@ list(APPEND QEC_SOURCES
...
+
+if(NOT TARGET libstim)
+  message(FATAL_ERROR
+    "Stim FetchContent did not provide the libstim target.")
+endif()
+target_link_libraries(${LIBRARY_NAME} PRIVATE libstim)
+target_link_options(${LIBRARY_NAME} PRIVATE
+  $<$<OR:$<CXX_COMPILER_ID:GNU>,$<CXX_COMPILER_ID:Clang>>:-Wl,--exclude-libs,libstim.a>
+)
+
```
Comment (bmhowe23) [inline libs/qec/lib/CMakeLists.txt:84]: I downloaded this PR and built locally (using 92d3dc977b46035cfd228ebf71e1b6cb524e9951), and I see the following, which could be problematic:

```bash
$ nm -D lib/libcudaq-qec.so | grep stim | c++filt
00000000000907d0 W stim::DetectorErrorModel::~DetectorErrorModel()
00000000000907d0 W stim::DetectorErrorModel::~DetectorErrorModel()
```

The above two items are the only problematic items, so it's possible that your solution here covers most of the issue, but I don't think we should be leaking those symbol out.

In order to prevent something like https://github.com/NVIDIA/cuda-quantum/issues/3045, I think something like https://github.com/NVIDIA/cuda-quantum/pull/3076 is required.

File: libs/qec/unittests/test_decoders.cpp [hunk 5]
```diff
@@ -743,63 +745,136 @@ TEST(DecoderTest, GetBlockSizeAndSyndromeSize) {
   EXPECT_EQ(decoder2->get_syndrome_size(), new_syndrome_size);
 }
 
-TEST(DecoderRegistryTest, SingleParameterRegistryDirect) {
```
Comment (bmhowe23) [inline libs/qec/unittests/test_decoders.cpp:746]: Just checking - does constructing a decoder w/o heterogenous map no longer supported? I don't think this is very common, so it's probably ok if we deprecate it; I just wanted to double check because if it still works then I'd want to keep this test.

**Review by bmhowe23 — COMMENTED** [review]
This is looking good...thanks, @vedika-saravanan. Just a few follow-up questions below.

File: libs/qec/unittests/test_decoders.cpp (hunk 5 above)
Comment (vedika-saravanan) [reply libs/qec/unittests/test_decoders.cpp:746]: Yes, constructing through the public API without an explicit heterogeneous_map is still supported; the overloads default to an empty map. The removed test was exercising the old single-argument extension_point registry directly, which no decoder registered into and which passed by expecting a runtime_error. I renamed the existing public-path test to make this coverage explicit: `DecoderTest.GetWithoutOptionsSetsBlockAndSyndromeSize` ([7b30f08](https://github.com/NVIDIA/cudaqx/pull/571/commits/7b30f089f03a08b27678816de16c7ad35b245e03)).

File: libs/qec/lib/CMakeLists.txt (hunk 4 above)
Comment (vedika-saravanan) [reply libs/qec/lib/CMakeLists.txt:84]: Thanks for the feedback @bmhowe23. I added `VISIBILITY_INLINES_HIDDEN ON` for `cudaq-qec` and rebuilt locally. The symbol check now returns no Stim exports:. Fixed in [7b30f08](https://github.com/NVIDIA/cudaqx/pull/571/commits/7b30f089f03a08b27678816de16c7ad35b245e03)

**Review by bmhowe23 — APPROVED** [review]
Thanks! Could you please create a follow-up PR/issue to add docs for this capability? We should add a public example for this.

**Comment by vedika-saravanan** [comment]
GHI for adding docs and example https://github.com/NVIDIA/cudaqx/issues/595

**Comment by bmhowe23** [comment]
@vedika-saravanan can you please do `ninja zip_installed_files` and inspect the zip file that it produces? I am seeing many stim binaries and directories that should not be "installed", and I suspect this PR is the one that added them.
